File: Medium/Evaluate Reverse Polish Notation/evaluate-reverse-polish-notation.cpp
class Solution {
public:
    int evalRPN(vector<string>& tokens) {
        
        vector<string> valid_ops = { "+", "-", "*", "/" };
        map<string, std::function<long(long,long)>> ops;
        ops.emplace("+", [=](long a, long b) { return a + b; });
        ops.emplace("-", [=](long a, long b) { return a - b; });
        ops.emplace("*", [=](long a, long b) { return a * b; });
        ops.emplace("/", [=](long a, long b) { return a / b; });

        stack<int> stack;

        for (auto t : tokens)
        {
            if (find(valid_ops.begin(), valid_ops.end(), t) != valid_ops.end())
            {
                int a = stack.top();
                stack.pop();
                int b = stack.top();
                stack.pop();
                stack.push(ops[t](b, a));
            }
            else
            {
                stack.push(stoi(t));
            }
        }

        return stack.top();
    }
};

File: Medium/Evaluate Reverse Polish Notation/evaluate-reverse-polish-notation.py
class Solution:
    def evalRPN(self, tokens: List[str]) -> int:

        vals = []
        available_ops = {
             "+": 0,
             "-": 0,
             "*": 1,
             "/": 1
        }

        ops_funcs = {
            "+": lambda x,y: x + y,
            "-": lambda x,y: x - y,
            "*": lambda x,y: x * y,
            "/": lambda x,y: int(x / y)
        }

        for t in tokens:

            if t in available_ops:
                a = vals.pop()
                b = vals.pop()
                vals.append(
                    ops_funcs[t](int(b), int(a))
                )                
            else:
                vals.append(t)

        return vals.pop()
